engine/runner/src/main/java/org/enso/runner/Main.java

```diff
@@ -1295,30 +1299,61 @@ private static <A> A withProfiling(
       java.util.concurrent.Callable<A> main)
       throws IOException {
     var path = profilingConfig.profilingPath();
+    var events = profilingConfig.profilingEventsLogPath();
     var sampler =
-        path.isDefined() ? OutputStreamSampler.ofFile(path.get().toFile()) : new NoopSampler();
+        path.isDefined() && events.isDefined()
+            ? MethodsSampler.create(path.get().toFile(), events.get().toFile())
+            : MethodsSampler.NOOP;
     sampler.start();
-    profilingConfig
-        .profilingTime()
-        .foreach(timeout -> sampler.scheduleStop(timeout.length(), timeout.unit(), executor));
+    profilingConfig.profilingTime().foreach(timeout -> sampler.scheduleStop(timeout));
     scala.sys.package$.MODULE$.addShutdownHook(
         () -> {
           try {
-            sampler.stop();
+            sampler.close();
           } catch (IOException ex) {
             LOGGER.error("Error stopping sampler", ex);
           }
           return BoxedUnit.UNIT;
         });
 
+    AutoCloseable c1 = null;
+    AutoCloseable c2 = null;
+    if (events.isDefined()) {
+      var eventsOut = new FileWriter(events.get().toFile());
+      eventsOut.write("<?xml version='1.0'?>\n");
+      eventsOut.write("<records>\n");
+      c1 = eventsOut;
+      var format = new XMLFormatter();
```

Comment:
I will probably move the XML dumping stuff directly to `MethodsSampler.log` message.